shuup/admin/modules/contacts/views/edit.py

```diff
@@ -32,144 +30,132 @@
+    "Contact": (
+        "is_active", "language", "marketing_permission", "phone", "www",
+        "timezone", "prefix", "suffix", "name_ext", "email", "tax_group",
+        "merchant_notes", "account_manager"
+    ),
+    "PersonContact": (
+        "first_name", "last_name", "gender", "birth_date"
+    ),
+    "CompanyContact": (
+        "name", "tax_number", "members"
+    )
+}
+
+FIELDS_BY_MODE = {
+    "create": {
+        "PersonContact": ("first_name", "last_name"),
+        "CompanyContact": FIELDS_BY_MODEL_NAME["CompanyContact"]
+    },
+    "update": {
+        "PersonContact": (list(FIELDS_BY_MODEL_NAME["PersonContact"]) + list(FIELDS_BY_MODEL_NAME["Contact"])),
+        "CompanyContact": (list(FIELDS_BY_MODEL_NAME["CompanyContact"]) + list(FIELDS_BY_MODEL_NAME["Contact"]))
     }
+}
+
 
-    def __init__(self, bind_user=None, *args, **kwargs):
-        class Meta:  # faux ModelForm Meta
-            model = Contact
-            fields = ()
+def get_contact_base_form(model_class, mode):
+    model_class_name = str(model_class.__name__)
+
+    class ContactBaseForm(forms.ModelForm):
+        class Meta:
+            model = model_class
+            fields = FIELDS_BY_MODE[mode][model_class_name]
             exclude = ()
 
-        self.base_fields = {}
-        self._meta = Meta
-        super(ContactBaseForm, self).__init__(*args, **kwargs)
-        self.contact_class = None
-        self.bind_user = bind_user
-
-        if self.bind_user:
-            self.contact_class = PersonContact
-            self.initial.setdefault("name", self.bind_user.get_full_name())
-        if self.instance.pk:
-            self.contact_class = self.instance.__class__
-
-        self.generate_fields()
-
-    def formfield_callback(self, f, **kwargs):
-        if str(f) == "shuup.CompanyContact.members":
-            formfield = Select2MultipleField(
-                model=PersonContact, required=False, help_text=f.help_text)
-            if hasattr(self.instance, "members"):
-                formfield.widget.choices = [
+        def __init__(self, *args, **kwargs):
+            super(ContactBaseForm, self).__init__(*args, **kwargs)
+            self.init_fields()
+
+        def init_fields(self):
+            self.fields["groups"] = forms.ModelMultipleChoiceField(
+                queryset=ContactGroup.objects.all(),
+                initial=(self.instance.groups.all() if self.instance.pk else ()),
+                required=False,
+                widget=forms.SelectMultiple(),
+                label=_("Contact Groups")
+            )
+            if "account_manager" in self.fields:
+                self.fields["account_manager"].widget = PersonContactChoiceWidget(clearable=True)
+
+        def save(self, commit=True):
+            obj = super(ContactBaseForm, self).save(commit)
+            obj.groups = self.cleaned_data["groups"]
+            return obj
+    return ContactBaseForm
+
+
+def get_person_contact_base_form(mode, user=None):
```

Comment:
I did this since I'm dynamically setting `fields` based on `mode`. I believe the alternative involves setting `Meta` inside of `__init__` which is just as messy imo. I suppose the easiest thing would be to just handle this `mode` logic inside the template. I will play around with this tomorrow...